# Welcome to the Police Station Lab Computer

Let's use the computer, and the data it contains, to see if we can't crack this mystery.

## First Visit

### Gathering Information on Addresses to interview the person who left before the gunshots were heard!

First thing to do, is load up the people `.csv` - we'll notice that it's separated by tabs (`\t`) and there's some comments (`#`), so let's take those into consideration. 

Then, let's look at the `.head()`!

### Summary of CLUES found in crimescene branch
    CLUE: Footage from an ATM security camera is blurry but shows that 
    - the perpetrator is a tall male, at least 6'.

    CLUE: Found a wallet believed to belong to the killer: 
    - no ID, 
    - just loose change, and 
    - membership cards for AAA, Delta SkyMiles, the local library, and the Museum of Bash History. 
    - The cards are totally untraceable and have no name, for some reason.

    CLUE: Questioned the barista at the local coffee shop. 
    - He said a woman left right before they heard the shots. 
    - The name on her latte was Annabel, 
    - she had blond spiky hair and a New Zealand accent.

### Objective:  interview the person who left before the gunshots

In [6]:
#--- check pwd
!pwd

/home/kidcoconut/myLurnins/fourthbrain.ai/code/whodunit


In [2]:
#--- ensure we are in the root project folder
%cd ..

/home/kidcoconut/myLurnins/fourthbrain.ai/code/whodunit


In [7]:
import pandas as pd

people_dataframe = pd.read_csv('./data/people/people.csv', comment='#', sep='\t')
people_dataframe.head()

name gender  age                    address
0  Alicia Fuentes      F   48    Walton Street, line 433
1   Jo-Ting Losev      F   46  Hemenway Street, line 390
2   Elena Edmonds      F   58   Elmwood Avenue, line 123
3  Naydene Cabral      F   46  Winthrop Street, line 454
4  Dato Rosengren      M   22    Mystic Street, line 477

Let's use some indexing to find out if we can narrow this list down!

In [8]:
### START CODE HERE ###

# be sure to include the information in the format provided from the DataFrame above!
#--- find all interviewees where name contains 'Annabel' and gender is 'F'
potential_interviewees = people_dataframe[(people_dataframe['name'].str.contains('Annabel')) & (people_dataframe['gender'] == 'F')]

### END CODE HERE ###

potential_interviewees.head()

name gender  age                     address
269      Annabel Sun      F   26         Hart Place, line 40
1661  Annabel Church      F   38  Buckingham Place, line 179

### Time for more investigation!

You'll need to hit the streets before you can pull up any relevant interviews, come back when you've knocked on the doors on the addresses you found!



    #--- Strategy:  I need to be able to manage the notebook, and data in one branch ... main


    #--- Police-Station Prereqs:  for analyzing data, and updating jupyter notebook in one place ...
    - checked out the police-station branch;  merged the notebook into main/notebooks


    #--- Streets Prereqs:  
    - checkout the streets branch;  merge addresses into main/data/streets
    
    
    #--- Streets Objective:  hit the streets;  visit the addresses found
    - cd ./data/streets
    - head -h | tail -t
    - head -40 Hart_* | tail -1:  SEE INTERVIEW #47246024
    - head -179 Buckingham_* | tail -1:  SEE INTERVIEW #699607


    #--- Interview Prereqs:
    - checked out the interviews branch;  merged into main/data/interviews


## Second Visit

### Checking out the interviews

Now that you have the interviews you want to check out, let's see what they say!

We'll use the `glob`, and `os` standard library to collect all the data together into a usable format

In [8]:
pwd

'/home/kidcoconut/myLurnins/fourthbrain.ai/code/whodunit'

In [13]:
import os 
import glob

strPattern = os.path.join('./data/interviews/', '*')
print("TRACE:  got file match pattern for glob;  ", strPattern, "\n")


aryFiles = glob.glob(strPattern)
#print(aryFiles)


TRACE:  got file match pattern for glob;   ./data/interviews/* 



Now that we've collected the files neatly, let's comb through them looking for the interviews we want to check!

In [28]:
### START CODE HERE ###

interview_nums = [
    "47246024", "699607",
]

### END CODE HERE ###

for strFile in aryFiles:
    strInterviewNum = strFile.split('-')[1]
    #print(strInterviewNum)
    if strInterviewNum in interview_nums:
        with open(strFile, 'r') as filTemp:
            print('---- Interview # {', strInterviewNum, '} ----')
            print(filTemp.read())
            print('\n')

---- Interview # { 699607 } ----
Interviewed Ms. Church at 2:04 pm.  Witness stated that she did not see anyone she could identify as the shooter, that she ran away as soon as the shots were fired.

However, she reports seeing the car that fled the scene.  Describes it as a blue Honda, with a license plate that starts with "L337" and ends with "9"



---- Interview # { 47246024 } ----
Ms. Sun has brown hair and is not from New Zealand.  Not the witness from the cafe.





Let's take note of that information, and check the vehicle database to see if we can find a suspect!

### Checking the Vehicle Database

Now that we have some more information, we should be able to put the facts we've figured out thusfar together to get a suspect!

Let's start by looking through the vehicle data!


    #--- Vehicle Prereqs: ...
    - checked out the police-station branch;  merge data/vehicles/vehicles.csv into main
    

In [9]:
### START CODE HERE ###

vehicle_dataframe = pd.read_csv('./data/vehicles/vehicles.csv', index_col=0)

### END CODE HERE ###

vehicle_dataframe.head()

License Plate      Make   Color              Owner Owner Height  \
0       T3YUHF6    Toyota  Yellow    Jianbo Megannem         5'6"   
1       EZ21ECE       BMW    Gold   Norbert Feldwehr         5'3"   
2       CQN2TJE     Mazda     Red  Alexandra Jokinen        5'11"   
3       D875IMS  Cadillac  Orange     Thi Kostadinov        5'11"   
4       Q0SK1KP  Cadillac     Red          Sidni Sze         5'9"   

   Owner Weight (lbs.)  
0                  246  
1                  205  
2                  227  
3                  198  
4                  199

We know the make and colour the suspect drove while leaving Pandas Express, thanks to the interview, and we also know the suspects height! We also know some information about the suspects license plate! 

Let's combine all this information and see what we can learn!

In [10]:
### START CODE HERE ###

# be sure to include the information in the format provided from the DataFrame above!
vehicle_suspects_dataframe = vehicle_dataframe[
    (vehicle_dataframe['Make'] == 'Honda') & 
    (vehicle_dataframe['Color'] == 'Blue') & 
    (vehicle_dataframe['License Plate'].str.startswith('L337')) &
    (vehicle_dataframe['License Plate'].str.endswith('9')) &
    (vehicle_dataframe['Owner Height'].str.startswith('6'))
]

### END CODE HERE ###

vehicle_suspects_dataframe.head()

License Plate   Make Color          Owner Owner Height  \
1023       L337QE9  Honda  Blue    Erika Owens         6'5"   
3547       L337DV9  Honda  Blue   Joe Germuska         6'2"   
4546       L3375A9  Honda  Blue  Jeremy Bowers         6'1"   
4901       L337WR9  Honda  Blue   Jacqui Maher         6'2"   

      Owner Weight (lbs.)  
1023                  220  
3547                  164  
4546                  204  
4901                  130

There we have it, a narrowed list of names - let's see if we can fit the final clue into the picture in order to wrap this case up!

### Gathering Information on Memberships

We have data from the following memberships on file:

1. `AAA`
2. `AAdvantage`
3. `Costco`
4. `Delta_SkyMiles`
5. `Fitness_Galaxy`
6. `Museum_of_Bash_History`
7. `REI`
8. `Rotary_Club`
9. `TCSU_Alumni_Association`
10. `Terminal_City_Library`
11. `United_MileagePlus`

From the evidence collected, input the memberships that the suspect held.

In [11]:
### START CODE HERE ###

# Be sure to include all of the suspects memberships in the format provided above!
memberships = [
    "AAA", "Delta_SkyMiles", "Terminal_City_Library", "Museum_of_Bash_History"
]

### END CODE HERE ###

memberships = list(map(lambda x: f"{x}.csv", memberships))

We're going to be using a combination of `glob` and `os` again! Because those libraries should already be imported, we doon't need to import them again!

In [14]:
files = os.path.join('./data/memberships/', '*.csv')
files = glob.glob(files)
print(files)

['./data/memberships/TCSU_Alumni_Association.csv', './data/memberships/Delta_SkyMiles.csv', './data/memberships/United_MileagePlus.csv', './data/memberships/Museum_of_Bash_History.csv', './data/memberships/Rotary_Club.csv', './data/memberships/Costco.csv', './data/memberships/Fitness_Galaxy.csv', './data/memberships/REI.csv', './data/memberships/AAdvantage.csv', './data/memberships/AAA.csv', './data/memberships/Terminal_City_Library.csv']


Next we'll combine these `.csv` files together into a list of `pandas.DataFrame`s (read more about those [here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html))

In [17]:
membership_dataframe_list = [pd.read_csv(file) for file in files if file.endswith(tuple(memberships))]
print(membership_dataframe_list)

[                   names
0           Ana Williams
1         Alejandro Abdi
2              Ana Dukic
3       Heather Billings
4       Lucia Maksimovic
...                  ...
1282   Alexandra Jokinen
1283      Stefan Webster
1284      Jemma Gabriele
1285  Alberto Fiakaifonu
1286        Mateusz Metu

[1287 rows x 1 columns],                    names
0          Limei Ioannou
1         Mihaela Toskic
2        Mihyun Dahlberg
3         Joseph Nielsen
4        Glenn Ouedraogo
...                  ...
1285     Emanuele Oulmou
1286         Peter Sloma
1287       Angela Palomo
1288     Peter Neethling
1289  Michelle Kirilenko

[1290 rows x 1 columns],                    names
0        Courtney Yankey
1     Robert Mothersille
2       Akvile Saedeleer
3         Daniel Hayashi
4            Roel Garcia
...                  ...
1299        Seiya Railey
1300     Radoslaw Robles
1301    Kathleen Schmidt
1302         Dani Loukas
1303        Kaspar Brown

[1304 rows x 1 columns],                      

Now, we'll find the intersection of the memberships using `pandas.merge` (read about it [here](https://pandas.pydata.org/docs/reference/api/pandas.merge.html)) on the list of dataframes.

We're using `reduce` from functools (read about it [here](https://www.geeksforgeeks.org/reduce-in-python/)) to perform the `pandas.merge` operation on a number of `pandas.DataFrame`s.

In [18]:
from functools import reduce

membership_suspects_dataframe = reduce(lambda x, y: pd.merge(x, y, how='inner'), membership_dataframe_list)
print(membership_suspects_dataframe)

                  names
0          Aldo Nicolas
1       Nikolaus Milatz
2              Emma Wei
3         Tamara Cafaro
4           Sonata Raif
5       Deron Estanguet
6          Andrei Masna
7         Marina Murphy
8          Krystian Pen
9        Mary Tomashova
10           Matt Waite
11         Jacqui Maher
12    Liangliang Miller
13         Jamila Rodhe
14         Mike Bostock
15          Brian Boyer
16         Didier Munoz
17      Augustin Lozano
18        Dalibor Vidal
19         Monika Hwang
20         Kelly Kulish
21  Stephanie Adlington
22        Jeremy Bowers


### Putting it all together

Alright, now that we have our vehicle suspects, our membership suspects, as well as the database of people - let's put this all together and see what we can find!

First step is to merge the two suspect lists to narrow it down.

In [19]:
merged_suspect_dataframe = pd.merge(vehicle_suspects_dataframe, membership_suspects_dataframe, left_on='Owner', right_on='names')
merged_suspect_dataframe.head()

License Plate   Make Color          Owner Owner Height  Owner Weight (lbs.)  \
0       L3375A9  Honda  Blue  Jeremy Bowers         6'1"                  204   
1       L337WR9  Honda  Blue   Jacqui Maher         6'2"                  130   

           names  
0  Jeremy Bowers  
1   Jacqui Maher

Gah! We have two suspects, let's check `people_dataframe` and see if there's anyway we can decide! 

HINT: Look at the new columns added as a result of the merge and remember what we know about our suspect!

In [20]:
final_merged_dataframe = pd.merge(merged_suspect_dataframe, people_dataframe, left_on="Owner", right_on='name')
final_merged_dataframe.head()

License Plate   Make Color          Owner Owner Height  Owner Weight (lbs.)  \
0       L3375A9  Honda  Blue  Jeremy Bowers         6'1"                  204   
1       L337WR9  Honda  Blue   Jacqui Maher         6'2"                  130   

           names           name gender  age                   address  
0  Jeremy Bowers  Jeremy Bowers      M   34  Dunstable Road, line 284  
1   Jacqui Maher   Jacqui Maher      F   40    Andover Road, line 224

HINT: Look at the new columns added as a result of the merge and remember what we know about our suspect! Pay close attention to the `gender` column!

Now that we have our final suspect, let's see if it's correct!

In [25]:
from helper_functions.check_name import check_name

check_name("Jeremy Bowers")

FileNotFoundError: [Errno 2] No such file or directory: '.env'